client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java

```diff
@@ -321,7 +350,36 @@ public List<AddBlockEvent> buildBlockEvents(List<ShuffleBlockInfo> shuffleBlockI
 
   @Override
   public long spill(long size, MemoryConsumer trigger) {
-    return 0L;
+    if (!memorySpillEnabled) {
+      return 0L;
+    }
+
+    // If it is triggered by other consumers, it will spilled async by current consumer.
+    if (trigger != this) {
+      spillTriggered.set(true);
+      return 0L;
+    }
+
+    List<CompletableFuture<Long>> futures = spillFunc.apply(clear());
```

Comment:
You will call the method `clear` in different threads. Is it thread safe?

Reply:
If I'm correct, I think you don't understand this design. Please see the below explaintion.